File: 0872-leaf-similar-trees/0872-leaf-similar-trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        return traverse(root1).equals(traverse(root2)) ;
    }
    
    String traverse(TreeNode root){
        if(root == null) return "";
        if(root.left ==null && root.right ==null) return root.val+"_"; 
        return traverse(root.left) + traverse(root.right ) ;
    }
}

File: 0872-leaf-similar-trees/0872-leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:


        def bfs(root): 


            st = [root] 
            ans = [] 
            while st: 
                node = st.pop() 
                if node.left: 
                    st.append(node.left)
                if node.right: 
                    st.append(node.right) 
                if not node.left and not node.right: 
                    ans.append(node.val) 

            return ans 

        return bfs(root1) == bfs(root2)
